test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,441 @@
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/gardener/etcd-druid/api/v1alpha1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/test/matchers"
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/coordination/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Etcd", func() {
+	Context("when multi-node is configured", func() {
+		var (
+			cl               client.Client
+			etcdName         string
+			storageContainer string
+			parentCtx        context.Context
+			provider         TestProvider
+			providers        []TestProvider
+			err              error
+		)
+
+		BeforeEach(func() {
+			parentCtx = context.Background()
+			providers, err = getProviders()
+			Expect(err).ToNot(HaveOccurred())
+			// take first provider
+			provider = providers[0]
+			cl, err = getKubernetesClient(kubeconfigPath)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			etcdName = fmt.Sprintf("etcd-%s", provider.Name)
+
+			storageContainer = getEnvAndExpectNoError(envStorageContainer)
+
+			snapstoreProvider := provider.Storage.Provider
+			store, err := getSnapstore(string(snapstoreProvider), storageContainer, storePrefix)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// purge any existing backups in bucket
+			Expect(purgeSnapstore(store)).To(Succeed())
+			Expect(deployBackupSecret(parentCtx, cl, logger, provider, etcdNamespace, storageContainer))
+
+		})
+
+		AfterEach(func() {
+			// remove etcd objects if any old etcd objects exists.
+			purgeEtcd(parentCtx, cl, providers)
+		})
+
+		It("should perform etcd operations", func() {
```

Comment:
Would it not make sense to use `BeforeAll`, `AfterAll` and `Ordered` containers?
One test here is doing way too much. I can see that to circumvent it you use `By` clause.
I generally prefer creating smaller tests where i can see all assertions and then individual test will only assert one behavior. Since it is ordered, you can use the state from the previous test as well.

If you see this one single BIG test then it tests for hibernation and wake up from hibernation, then it also tests  rolling update, defragmentation etc. These are not really correlated.

Reply:
> Since it is ordered, you can use the state from the previous test as well.

Can you explain what you mean by ordered? There is an order decorator in Ginkgo 2.x which we currently don't use [ref](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#ordered-decorator). Otherwise there is no order of `It`s as far as I can see 👀

Reply:
Yes [this](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#ordered-decorator) is correct. I see that etcd-druid currently only uses v1 - :(